runtime/src/chpl-gpu.c

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2020-2021 Hewlett Packard Enterprise Development LP
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.  * 
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * 
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// This #define needs to be before the other #includes
+// since it affects included files
+#ifdef __linux__
+#define _GNU_SOURCE
+#endif
+
+#include "chplrt.h"
+#include "chplsys.h"
+
+#ifdef HAS_GPU_LOCALE
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <assert.h>
+
+static void checkCudaErrors(CUresult err) {
+  assert(err == CUDA_SUCCESS);
```

Comment:
This won't do anything in a release build. Is that what you want? Also this is only called in 1 place. Are the other CUDA calls ones that can fail? Should we check them?

`runtime/include/error.h` has some functions you can call for fatal errors.

Reply:
This is code that I factored out from the test, but it does raise some interesting questions about how we should address CUDA errors.

Apretty common pattern I've seen in CUDA code is to wrap all the calls inside a macro or a function like this that asserts they succeeded. I think it makes sense for us to follow this pattern, so I've updated all the cuda calls to be wrapped in this function.

I've also updated the function to use chpl_internal_error. I think eventually we'll want to give more eloquent errors for the sort of circumstances that could occur that arguably wouldn't be caused due to internal problems with the Chapel compiler.

For example if someone tries to run this program on a machine that doesn't have a GPU or if they try and allocate more memory on the GPU than is available we'd want to present a better error message then just saying it's an "internal error calling out to a CUDA function".

But I'll leave that as future work for now.